### PR TITLE
adding a function based dispatch.

### PR DESCRIPTION
**Pull request by toumorokoshi**
This address #9.

A proof of concept PR, so I could do more polish on it. But it illustrates:

1. using a FunctionDispatch class vs SingleDispatch.
2. feature parity with SingleDispatch (including caching for quick lookup)

Tried it out locally, works with py27 and py36, sans the vendor typing library.

More improvements / changes:

- change the register function signature to take two functions, rather than type + function.
- use FunctionDispatch to also handle attrs

### CODE REVIEW

**Comment by Tinche** [comment]
Thanks! I'll take a look soon.

**Comment by Tinche** [comment]
The problem with this approach is that the following snippet stops working:
```
import attr
import cattr


@attr.s
class A:
    a = attr.ib()


@attr.s
class B(A):
    b = attr.ib()


cattr.register_structure_hook(B, (lambda d, t: B(**d)))
cattr.register_structure_hook(A, (lambda d, t: A(**d)))

print(cattr.structure({'a': 1, 'b': 2}, B))
```

I'm OK with having a list of matchers that are consulted in turn when looking up a handler function (and we will probably require this to handle Unions with parameters), but the last entry in this list should be something that understands the class hierarchy and the MRO. Then the API that prepends to the list would be a low level kind of API, and just registering a type in the last entry would be a high level kind of API.

**Comment by toumorokoshi** [comment]
Yeah, that approach sounds reasonable: best of both worlds. 

Probably will work with singledispatch at the end, with a caveat on it's use on Typing objects. I'll give it a try.

**Comment by toumorokoshi** [comment]
I'll add some more coverage for the hooks, but my general approach is done. I'll probably start commenting now.

**Comment by Tinche** [comment]
Thanks for your effort. I will do my best to review over the weekend.

File: tests/test_multistrategy_dispatch.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+def _foo_func():
+    pass
+
+
+def _foo_cls():
+    pass
+
+
+def test_multistrategy_dispatch_register_cls():
+    _fallback()
```
Comment (toumorokoshi) [inline tests/test_multistrategy_dispatch.py]: If I didn't add these calls, the coverage wasn't sufficient. 

I'm not sure how to tackle that, so I just call them.

File: tests/test_structure.py [hunk 2]
```diff
@@ -279,3 +301,23 @@ def test_structuring_unsupported(converter):
         converter.structure(1, Converter)
     with raises(ValueError):
         converter.structure(1, Union[int, unicode])
+
+
+def test_subclass_registration_is_honored(converter):
```
Comment (toumorokoshi) [inline tests/test_structure.py]: A unit test for the behavior you brought up in the PR.

File: cattr/multistrategy_dispatch.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    """ a dummy object to help signify a dispatch not found """
+    pass
+
+
+class MultiStrategyDispatch(object):
+    """
+    MultiStrategyDispatch uses a
+    combination of FunctionDispatch and singledispatch.
+
+    singledispatch is attempted first. If nothing is
```
Comment (toumorokoshi) [inline cattr/multistrategy_dispatch.py]: Just wanted to note here: we discussed using FunctionDispatch, then falling back to SingleDispatch, but I think that won't work to resolve subclasses: registering a subclass of an iterable, then having the functiondispatch pick that up when determined as a subclass for typing.List and runs iterable_seq.

So switched the order: singledispatch is attempted, then FunctionDispatch as a fallback.

**Comment by Tinche** [comment]
Looking good! Almost ready for merging.

Some comments/questions:
* Can we remove the vendored typing for Python 2?
* Some classes should be converted to attrs classes :) (FunctionDispatch and _DispatchNotFound for sure.)
* Please put in a line in HISTORY.rst about the change.

**Comment by toumorokoshi** [comment]
Ok, addressed 2 and 3.

For 1: I didn't want to pollute this PR with a bunch of changes on the import statements, so I created another pr (#12).

**Comment by Tinche** [comment]
Thank you for your contributions!

Just FYI, I'll probably change the API for registering hooks a little before release (by adding a special call for generics). Also I don't plan on releasing until attrs 17.3 is out because then we can remove cattr.typed in favor of attr.ib(type=).

**Comment by toumorokoshi** [comment]
yep, both of those makes sense. I'd still like to have some way to register a function hook publicly though: that will enable me to register non-class based systems like schematics.

Thanks!

**Pull request merged**